treeherder/auth/backends.py

```diff
@@ -0,0 +1,148 @@
+                # use their clientId as a dummy email.  We don't
+                # send anything to it, so it doesn't matter much, in
+                # reality.  This can only happen if the user isn't using LDAP,
+                # and doesn't have an email in their clientId.
+                email = result["clientId"]
+        return email
+
+    def _get_user(self, username, email):
+        """
+        Try to find an exising user that matches either the username
+        or email.  Prefer the username, since that's the unique key.  But
+        fallback to the email.
+
+        """
+
+        # Since there is a unique index on username, but not on email,
+        # it is POSSIBLE there could be two users with the same email and
+        # different usernames.  Not very likely, but this is safer.
+        user = User.objects.filter(
+            Q(username=username) | Q(email=email))
+
+        # if we didn't find either, then raise an exception so we create a new
+        # user
+        if not len(user):
+            raise ObjectDoesNotExist
+
+        # prefer a matching username.  If no username match, return the first
+        # user in the list.
+        try:
+            return user.get(username=username)
+        except ObjectDoesNotExist:
+            return user.first()
+
+    def authenticate(self, authorization=None, host=None, port=None):
+        user = None
+
+        if authorization:
+            tc_auth = Auth()
+            result = tc_auth.authenticateHawk({
+                "authorization": authorization,
+                "host": host,
+                "port": port,
+                "resource": "/",
+                "method": "get",
+            })
+
+            if result["status"] == "auth-success":
+                # TODO: remove this size limit when we upgrade to django 1.10
+                # in Bug 1311967
+                username = result["clientId"][-30:]
+                email = self._get_email(result)
+
+                try:
+                    # Either find the user by their username or email.
+                    user = self._get_user(username, email)
+
+                except ObjectDoesNotExist:
+                    # the user doesn't already exist, create it.
+                    logger.warning("Creating new user: {}".format(username))
+                    sha = sha1()
+                    sha.update(email)
+                    is_staff = scope_match(
+                        result["scopes"],
+                        [["assume:project:treeherder:sheriff"]]
+                    )
+                    user = User(email=email,
+                                username=username,
+                                password=sha.hexdigest()[-25:],
+                                is_staff=is_staff
+                                )
+
+                # update their email if their scopes when the user was created
+                # didn't allow finding the email, but now it's been fixed.
```

Comment:
I think this comment is now out of date?

Do we want to update their username (away from the current hash value) instead?